### PR TITLE
Grade-mini - a CRT emulation shader

### PR DESCRIPTION
**Pull request by Dogway**
This is an abridged port of Grade, a popular (initially grading only) shader for RetroArch and now included in crt-guest-advanced and most preset packages like the Mega Bezel Reflection Shader, CyberLab Mega Bezel, Sonkun’s crt-guest-advanced-hd presets, etc.

I started development in 2020, over 3 years ago and in the last two months I updated it to include gamut compression, a faithful model of a CRT transfer function and a more accurate white point mapping function.

A very old (and primitive) version of the white point mapping function was also committed to PCSX2 to allow the reproduction of Japanese developed games in their original cool temperature.

For this version, called grade-mini I stripped unrelated or highly specific settings to avoid confusion and make it more approachable.

In summary it emulates composite signal round trip (conversion to Y'UV and quantization), CRT transfer function, RGB CRT Beams, CRT Saturation, CRT phosphor gamut and color temperature. For the output it implements gamut compression and color space conversion.

This commit covers and expands on the following issue, opened by me under a different alias:
[https://bugs.dolphin-emu.org/issues/8941](https://bugs.dolphin-emu.org/issues/8941)

**EDIT**: Screenshots as requested

MGS 01
MGS 02
SMG 01
SMG 02
WW ORIG
NTSC-J phosphors (as above examples)
WW NTSC-J
CRT Projector phosphors
WW Proj
TP03 OFF PNG
NTSC-J phosphors
TP04 NTSC-J PNG
CRT Projector phosphors
TP05 Projector PNG

### CODE REVIEW

**Comment by iwubcode** [comment]
> *It works on Vulkan and OpenGL, but fails in DX11 due to a bug in the shader translation layer being unable to translate functions with mat3 types.

What was the error you saw?

**Comment by Dogway** [comment]
It simply ignores the shader. If you mean the log, here it is:

```
24:02:186 VideoBackends\D3DCommon\Shader.cpp:123 E[MASTER]: Warning: Failed to compile S:/Emulators/Nintendo Wii/dolphin-master-5.0-16796-x64/User/Dump/bad_ps_5_0_D3D_1.txt: Error no especificado (0x80004005)
Debug info (ps_5_0):
S:\Emulators\Nintendo Wii\dolphin-master-5.0-16796-x64\Shader@0x000000000797DD80(250,1-4): error X3000: unrecognized identifier 'mat3'
24:02:186 VideoCommon\PostProcessing.cpp:743 E[MASTER]: Warning: Failed to compile post-processing shader grade-mini

24:02:843 VideoBackends\D3DCommon\Shader.cpp:130 W[Video]: ps_5_0 compilation succeeded with warnings:
S:\Emulators\Nintendo Wii\dolphin-master-5.0-16796-x64\Shader@0x000000000907F0A0(21,18-73): warning X3571: pow(f, e) will not work for negative f, use abs(f) or conditionally handle negative values if you expect them
```

The second is a warning that I will have a look on. The Dolphin version is a one year old one since I'm still on Win7, if you want to test on Windows10 that would be appreciated.

**Comment by JMC47** [comment]
I don't see any problems with adding new shaders, but the fact it doesn't work in D3D makes it hard to justify merging this.  I think that should be a top priority - making sure shaders we ship with Dolphin work with dolphin.

**Comment by Dentomologist** [comment]
Given that you're on 5.0-16796, the pow() warnings might be the ones fixed in [5.0-16831](https://dolphin-emu.org/download/dev/fac66897af7336b5e2dc2e3b879effb98b30fe47/).

**Comment by MayImilae** [comment]
You should post some screenshots!

**Comment by Dogway** [comment]
> You should post some screenshots!

Updated OP with some screenshots and also updated the shader with some clarification notes. Also reduced saturation a bit by default since typically saturation is lost through the signal in analogous systems.

**Comment by iwubcode** [comment]
Yes, I can only assume you are on a _very_ old Dolphin version from dev standpoints.  I am able to run the shader using the DirectX backend just fine.  I do see some errors reported in the console, for instance, I was able to see that `m_in` wasn't initialized.

I end up seeing something odd where blacks are all white on the DirectX backend (toggling the `Composite Signal Type` fixes it)... otherwise the shader seems to operate alright.

Whether we want this, I'll leave up to other contributors to decide.  I have no interest in CRT emulation personally.

**Comment by Dogway** [comment]
_Very_ or not, I don't know. The build is 11 months old. Which DX backend?

Fixed the 'm_in' initialization issue, as for the blacks turning white (NaN issue probably) I'm not sure where to look, it's happening on the first code block, that's for sure.

> Whether we want this, I'll leave up to other contributors to decide. I have no interest in CRT emulation personally, so I can't say if this is accurate or desired but it works fine.

Yeah, some people like their squares pixels too (think native resolution here), but as said in OP 'Grade' is a staple in all [RetroArch](https://forums.libretro.com/c/retroarch-additions/retroarch-shaders) shader presets for a few years now as well as PCSX2 (a basic white point adjustment).
Maybe we would need a poll for Dolphin's shader collection because I personally find no use for toxic, chrismas or acidtrip to name a few.

**Comment by AdmiralCurtiss** [comment]
@iwubcode So is this ready for merge? I was kinda ignoring this because of the mentioned failures on D3D but apparently that works fine?

**Comment by AdmiralCurtiss** [comment]
(well the commits need squashing but I mean, code-wise)

**Comment by Dogway** [comment]
Updated with some cosmetics.
Somebody has to confirm if the NaN issue is solved for D3D as I added a clamp() where I think it might had been failing.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by iwubcode** [comment]
> I was kinda ignoring this because of the mentioned failures on D3D but apparently that works fine?

From a high level, does this work standpoint, yes, it works.  The statement in the opening is incorrect as they are using a build from almost a _year ago_ to test.

But it still has the blacks are white issue that I mentioned earlier.

**Comment by Dogway** [comment]
Please check now on D3D.

**Comment by AdmiralCurtiss** [comment]
@iwubcode Can you check this again?

**Comment by AdmiralCurtiss** [comment]
white-d3d

Whites are still broken in D3D.

File: Data/Sys/Shaders/grade-mini.glsl [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+[OptionRangeInteger]
+GUIName = Phosphor (-3:170M -2:RPTV-95s -1:P22-80s 1:P22-90s 2:NTSC-J 3:PAL)
+OptionName = g_crtgamut
+MinValue = -3
+MaxValue = 3
+StepAmount = 1
+DefaultValue = 2
+
+[OptionRangeInteger]
+GUIName = Diplay Color Space (0:709 1:sRGB 2:P3-D65 3:Custom (Edit L508))
```
Comment (AdmiralCurtiss) [inline Data/Sys/Shaders/grade-mini.glsl:76]: There's a typo here, `Diplay`.

Comment (Dogway) [reply Data/Sys/Shaders/grade-mini.glsl:76]: @AdmiralCurtiss Thanks for feedback, can you check with 19435? I want to rule out the HDR thingy implemented after it is messing with the whites. And if possible without texture mods.
Otherwise I can only add tighter clamps until next visit to my friend's house.

**Comment by Dogway** [comment]
Now should be working for DX11, thank you.

**Comment by AdmiralCurtiss** [comment]
Also produces an error when loading in OpenGL:

grafik

[bad_p_OGL_1.txt](https://github.com/dolphin-emu/dolphin/files/12227224/bad_p_OGL_1.txt)

**Comment by Dogway** [comment]
> Still broken, unfortunately. Pure blacks now render as pink in DX11.
> 

What is that, doesn't DX11 really understand 0/255 is 0? do I have to spoonfeed it all?

Try the next, and if still fails I'll remove the whole clamp() which is what exactly was working at my friend's house.

**Comment by AdmiralCurtiss** [comment]
Nope, still wrong. I threw this into RenderDoc and you seem to be producing a `nan` (via `atan2(0, 0)`), which is not surprising that it behaves differently between backends (IIRC DirectX optimizes around assuming nans never happen).

grafik

**Comment by Dogway** [comment]
In the version I tested which is more recent but not latest , DX11 had no problems with atan () it was all about  the Quantize function. Then I added the clamp for yuv  so the conversion makes sense , nothing else . 

So all I need there is a ternary you say ?

**Comment by AdmiralCurtiss** [comment]
Yes, `float    hue  = col.z == 0.0 && col.y == 0.0 ? 0.0 : atan(col.z,  col.y);` fixes it. `atan2(0, 0)` is undefined so the fact that OpenGL/Vulkan worked fine was kind of a happenstance.

**Comment by Dogway** [comment]
> Yes, `float hue = col.z == 0.0 && col.y == 0.0 ? 0.0 : atan(col.z, col.y);` fixes it. `atan2(0, 0)` is undefined so the fact that OpenGL/Vulkan worked fine was kind of a happenstance.

Thank you, updated, I think only col.z check is necessary, tested this on OpenGL and everuthing seems to be working as expected so I hope DX11 does too.

**Comment by Dogway** [comment]
Was expecting to make it at least to the progress report so at least people have a sane alternative to the mess of https://github.com/dolphin-emu/dolphin/pull/11850

Is there any issue preventing this from being merged? This is taking longer than I expected and I need to move on.

**Comment by iwubcode** [comment]
You being stuck on a very old version of Dolphin and therefore not being able to do proper testing for D3D has been the main blocker.  If D3D is working, I see no problem merging it.

Also, not sure how your post processing shader would have any impact on what the user is complaining about there.  That has to do with the swapchain format impacting the driver's ability to engage GSYNC as far as I can tell.

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by Dogway** [comment]
> You being stuck on a very old version of Dolphin and therefore not being able to do proper testing (in particular for D3D) has been the main blocker. If D3D is working, I see no problem merging it.
> 
> Also, not sure how your post processing shader would have any impact on what the user is complaining about there. That has to do with the swapchain format impacting the driver's ability to engage GSYNC as far as I can tell.

The last "fixing" commit was from 7 days ago, anyway if you have followed the PR I've been testing on a friend's house with Windows10 and D3D gave no issues with a build just before the PR I linked. If that PR caused any issues as noted by AdmiralCurtis it would be dealt with the changes he suggested, so it might be related, so far it seems 10-bit framebuffer can't be disabled leading to (so far) unknown bugs.

**Comment by Dogway** [comment]
Retiring the PR due to lack of interest.

I start classes on Tuesday and have no more time to deal with apathy and disdain attitudes. First time I experience this on a PR, 3 months and ongoing to commit a mere non-disruptive and high quality shader tried-and-tested for 3 years.

For those that are still interested, follow the next links:

My Github repo for [Dolphin shaders](https://github.com/Dogway/emulation-random/tree/master/Dolphin/Shaders), includes grade-mini and PAR_Zoom_Offset.
The original Grade [thread](https://forums.libretro.com/t/dogways-grading-shader-slang/27148) in the libretro forums.

Thanks for everything.
